### ISSUE TITLE
Is it possible to get neighbour cell or points?

### ISSUE THREAD

**Issue by pswpswpsw** [akaszynski/vtkInterface#23]
I don't want to look deep into VTK but just simply using your vtkInterface to get the information for a cell or point, getting the index for nearby points.

For example, i can use `grid.cell_arrays` or `grid.point_array` together with `grid.points` to get a `dict` of all the "nearly useful" information. But I want to obtain the nearest neighbour for a given point, which enables much flexibility in postprocessing the vtk formatted data. Is it possible using the existing method in vtk? I am asking because I find something like `GetCellNeighbors` but since it requires going deep into vtk so I cound't do it on my own. 

Alternatively, I know I can using another package such as `[kdTree](http://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KDTree.html)` to search for `k-nearest neighour` on my own but just curious if there is something already there.

By the way, I think using the `kdTree` would enable much more flexibility since if I want to compute a very high order finite difference, I need **a large stencil** which I think  `GetCellNeighbors` won't give me but selecting a large enough `k-nearest neighour` might give me. I don't know if it is something also your interests.

**Comment by akaszynski** [comment]
I’m not familiar with what a large stencil is, but I’d probably go with using a kdtree as you’ll be able to quickly query the tree once you’ve set it up.  pykdtree is one of the best out there, and it carries no dependancies.

If you do need me to setup an adjency function, please comment with your intended pseudo-code.

**Comment by akaszynski** [comment]
Still need help on this?

**Comment by supersubscript** [comment]
I agree that it would be nice to have a few different adjacency/neighbourhood functions. These examples come to mind for neighbouring elements on a mesh:
- https://www.vtk.org/Wiki/VTK/Examples/Cxx/PolyData/VertexConnectivity
- https://www.vtk.org/Wiki/VTK/Examples/Cxx/PolyData/CellPointNeighbors
- https://www.paraview.org/Wiki/VTK/Examples/PolyData/CellEdgeNeighbors

And then this for a point cloud:
- https://www.vtk.org/Wiki/VTK/Examples/Cxx/DataStructures/KdTreePointLocator/ClosestNPoints

Would certainly be features that'd be useful to have.

**Comment by pswpswpsw** [comment]
> 
> 
> Still need help on this?

it is not urgent for me but I think other people might be interested. 
Btw I wrote a [blog](https://pswpswpsw.github.io/posts/2018/09/blog-post-modify-vtk-openfoam/) on postprocessing OpenFOAM data using VTK and then use your package for various purpose in Python: 

So the pseudo-code I have in my mind would be

-----
let's say we tried to find 2-level of Neighour for each cell
for loop over all the cells, say i-th cell for instance:
    - find the neighbour of i-th cell (I think VTK itself might have already done it for you), denote as set J_i
         for loop over each cell in set J_i, for each cell, find the neighour of that cell, denote as J_{J_i}. 
    - for each cell Let's do a union operation of all the set J_{J_i}. Then This union set for i-cell is the 2-level neighour set. Note that we know all of the spatial information so we can do really fancy things like computing any partial derivatives on any directions, so getting curvature is trivially easy.

**Comment by akaszynski** [comment]
This is different than a meshless kdtree as it requires knowledge of the mesh connectivity.  I’d be happy to add in cell based neighbor querying.

BTW, I liked the blog. If you wish, you can also visualize within python without heading over to paraview with:
```python
grid.Plot(scalars=p2_cell)
```

See help(grid.Plot) for more details on how you can change the color map and other plotting variables.

**Comment by banesullivan** [comment]
@pswpswpsw, could we add your blog post to the [Long Examples](
http://www.vtki.org/en/latest/examples/index.html#long-examples) section of the website?

Also, is there a status on the proposed feature requests? Should we outline some features that should/could be added to `vtki`?

**Comment by pswpswpsw** [comment]
> @pswpswpsw, could we add your blog post to the [Long Examples](http://www.vtki.org/en/latest/examples/index.html#long-examples) section of the website?
> 
> Also, is there a status on the proposed feature requests? Should we outline some features that should/could be added to `vtki`?

No problem! Indeed, I will have some more fancy figures later within 2 months on that post! I will keep you updated!
